url89849c8c5b_timmolter_XChart_tgz-pJ8-org_knowm_xchart_standalone_Example1J8/src/org/knowm/xchart/internal/chartpart/AxisTickLabels.java

```diff
@@ -0,0 +1,209 @@
+ * limitations under the License.
+ */
+package org.knowm.xchart.internal.chartpart;
+
+import java.awt.Graphics2D;
+import java.awt.Shape;
+import java.awt.font.FontRenderContext;
+import java.awt.font.TextLayout;
+import java.awt.geom.AffineTransform;
+import java.awt.geom.Rectangle2D;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.knowm.xchart.internal.Series;
+import org.knowm.xchart.internal.Series_AxesChart;
+import org.knowm.xchart.internal.chartpart.Axis.Direction;
+import org.knowm.xchart.style.AxesChartStyler;
+
+/**
+ * Axis tick labels
+ */
+public class AxisTickLabels<ST extends AxesChartStyler, S extends Series> implements ChartPart {
+
+  private final Chart<AxesChartStyler, Series_AxesChart> chart;
+  private Rectangle2D bounds;
+  private final Direction direction;
+
+  /**
+   * Constructor
+   *
+   * @param chart
+   * @param direction
+   */
+  protected AxisTickLabels(Chart<AxesChartStyler, Series_AxesChart> chart, Direction direction) {
+
+    this.chart = chart;
+    this.direction = direction;
+  }
+
+  @Override
+  public void paint(Graphics2D g) {
+
+    g.setFont(chart.getStyler().getAxisTickLabelsFont());
+
+    g.setColor(chart.getStyler().getAxisTickLabelsColor());
+
+    if (direction == Axis.Direction.Y && chart.getStyler().isYAxisTicksVisible()) { // Y-Axis
+
+      double xWidth = chart.getYAxis().getAxisTitle().getBounds().getWidth();
+      double xOffset = chart.getYAxis().getAxisTitle().getBounds().getX() + xWidth;
+      double yOffset = chart.getYAxis().getBounds().getY();
+      double height = chart.getYAxis().getBounds().getHeight();
+      double maxTickLabelWidth = 0;
+      Map<Double, TextLayout> axisLabelTextLayouts = new HashMap<Double, TextLayout>();
+
+      for (int i = 0; i < chart.getYAxis().getAxisTickCalculator().getTickLabels().size(); i++) {
+
+        String tickLabel = chart.getYAxis().getAxisTickCalculator().getTickLabels().get(i);
+        // System.out.println("** " + tickLabel);
+        double tickLocation = chart.getYAxis().getAxisTickCalculator().getTickLocations().get(i);
+        double flippedTickLocation = yOffset + height - tickLocation;
+
+        if (tickLabel != null && flippedTickLocation > yOffset && flippedTickLocation < yOffset + height) { // some are null for logarithmic axes
+          FontRenderContext frc = g.getFontRenderContext();
+          TextLayout axisLabelTextLayout = new TextLayout(tickLabel, chart.getStyler().getAxisTickLabelsFont(), frc);
+          Rectangle2D tickLabelBounds = axisLabelTextLayout.getBounds();
+          double boundWidth = tickLabelBounds.getWidth();
+          if (boundWidth > maxTickLabelWidth) {
+            maxTickLabelWidth = boundWidth;
+          }
+          axisLabelTextLayouts.put(tickLocation, axisLabelTextLayout);
+        }
+      }
+
+      for (Double tickLocation : axisLabelTextLayouts.keySet()) {
+
+        TextLayout axisLabelTextLayout = axisLabelTextLayouts.get(tickLocation);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.